drivers/tee/optee/call.c

```diff
@@ -244,32 +383,124 @@ int optee_invoke_func(struct tee_context *ctx, struct tee_ioctl_invoke_arg *arg,
 	if (!sess)
 		return -EINVAL;
 
-	shm = get_msg_arg(ctx, arg->num_params, &msg_arg, &msg_parg);
-	if (IS_ERR(shm))
-		return PTR_ERR(shm);
-	msg_arg->cmd = OPTEE_MSG_CMD_INVOKE_COMMAND;
-	msg_arg->func = arg->func;
-	msg_arg->session = arg->session;
-	msg_arg->cancel_id = arg->cancel_id;
+	ocall = tee_param_find_ocall(param, arg->num_params);
+	if (IS_ERR(ocall))
+		return PTR_ERR(ocall);
 
-	rc = optee_to_msg_param(msg_arg->params, arg->num_params, param);
-	if (rc)
-		goto out;
+	act_num_params = arg->num_params - (ocall ? 1 : 0);
+	act_params = param + (ocall ? 1 : 0);
 
-	if (optee_do_call_with_arg(ctx, msg_parg)) {
-		msg_arg->ret = TEEC_ERROR_COMMUNICATION;
-		msg_arg->ret_origin = TEEC_ORIGIN_COMMS;
+	/*
+	 * If OCALL requests are not expected, forgo the overhead of allocating
+	 * a full calling context and its associated elements.
+	 */
+	if (!ocall)
+		return optee_invoke_func_fast(ctx, arg, param);
+
+	if (!ctx->cap_ocall)
+		return -ENOTSUPP;
+
+	ocall_id = tee_param_get_ocall_id(ocall);
+	if (ocall_id) {
+		/*
+		 * The current call is a reply to an OCALL request.
+		 */
+
+		call_ctx = optee_ocall_ctx_get_from_id(ctx, ocall_id);
+		if (IS_ERR(call_ctx))
+			return PTR_ERR(call_ctx);
+
+		rc = optee_ocall_process_reply(arg, act_params, act_num_params,
+					       ocall, call_ctx);
+		if (rc)
+			goto exit;
+	} else {
+		/*
+		 * The current call is a regular function invocation that may
+		 * result in an OCALL request.
+		 */
+
+		call_ctx = optee_ocall_ctx_alloc(ctx, act_num_params,
```

Comment:
Your suggestion presumes that TAs will remain be single-threaded. It's true now, but we'd be interested in lifting that restriction at some point.

There is further trouble in that sessions aren't reference-counted. It would be possible for another thread to close and free the session while it's being used by a function invocation. I could of course add a reference count to sessions, assuming you'd be OK with that.

I added the fast path only to avoid the allocation of the context; it all works just the same without it, we don't need it per-se and I'm happy to do away with it.

I'd prefer leaving the call context separate from the session for future-proofing. Having said that. it's likely that if and when TAs gain the ability to support multiple concurrent threads, changes will be required to the driver, so the counter-argument would be to wait until then.

Thoughts?

Reply:
One of the threading models I'm considering is using one session per thread so this would be another argument for that model.

Threads (in the CA) are supposed to cooperate, so if one closes a session which has an ongoing OCALL it's the wish of the CA to cancel the current invoke and get rid of the session. What we need to worry about here is that the session/OCALL always is in a consistent state. Perhaps that's what you mean by reference counting? Yes, it's fine to make sure that only one "invoke" at a time is using a specific session.

My main concern about future-proofing is the two ABIs between user space and the driver, and between the driver and secure world. So far it seems we have that covered.

Reply:
> Threads (in the CA) are supposed to cooperate

Would this argument also apply to threads created by the TA (assuming we provide e.g., `pthread_create()` to TAs)?

Reply:
The short answer is yes, since they share resources that must be expected.
I'm happy to discuss multithreaded TAs further, but perhaps outside of this PR?

Reply:
Of course. In fact I think this OCALL mechanism could be a key enabler for MT TAs, so the discussion can wait until this is done. I just wanted to make sure no blocker would be introduced.

Reply:
> In fact I think this OCALL mechanism could be a key enabler for MT TAs...

Agree